### PR TITLE
oci: use json formatted errors from the runtime

### PR DESCRIPTION
**Pull request by giuseppe**
request json formatted error messages from the OCI runtime so that we
can nicely print them.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
it is would be safer though to do it only with runc, any idea that doesn't require to hard code the runtime name?

File: libpod/oci_linux.go [hunk 1]
```diff
@@ -417,6 +427,14 @@ func (r *OCIRuntime) createOCIContainer(ctr *Container, cgroupParent string, res
 		}
 		logrus.Debugf("Received container pid: %d", ss.si.Pid)
 		if ss.si.Pid == -1 {
+			data, err := ioutil.ReadFile(ociLog)
+			if err == nil {
+				var ociErr ociError
+				if err := json.Unmarshal(data, &ociErr); err == nil {
+					return errors.Wrapf(ErrInternal, "container create failed: %s", strings.Trim(ociErr.Msg, "\n"))
```
Comment (rhatdan) [inline libpod/oci_linux.go]: This should not be an ErrInternal, I think we should just return the strings.Trim(ociErr.Msg, "\n")
return errors.Errorf(strings.Trim(ociErr.Msg, "\n"))

File: libpod/oci_linux.go [hunk 2]
```diff
@@ -417,6 +427,14 @@ func (r *OCIRuntime) createOCIContainer(ctr *Container, cgroupParent string, res
...
+				}
+			}
+			// If we failed to parse the JSON errors, then print the output as it is
 			if ss.si.Message != "" {
 				return errors.Wrapf(ErrInternal, "container create failed: %s", ss.si.Message)
```
Comment (rhatdan) [inline libpod/oci_linux.go]: Here I also just return 
return errors.Errorf(ss.si.Message)
This is not an internal error and we did create the container,

Comment (giuseppe) [reply libpod/oci_linux.go]: fixed

File: libpod/oci_linux.go (hunk 1 above)
Comment (mheon) [reply libpod/oci_linux.go]: I'd add a new `ErrOCIRuntime` for errors in the OCI runtime and wrap that

File: libpod/oci_linux.go [hunk 3]
```diff
@@ -219,6 +228,7 @@ func (r *OCIRuntime) createOCIContainer(ctr *Container, cgroupParent string, res
 	args = append(args, "-b", ctr.bundlePath())
 	args = append(args, "-p", filepath.Join(ctr.state.RunDir, "pidfile"))
 	args = append(args, "--exit-dir", r.exitsDir)
+	args = append(args, "--runtime-arg", "--log-format=json", "--runtime-arg", "--log", fmt.Sprintf("--runtime-arg=%s", ociLog))
```
Comment (mheon) [inline libpod/oci_linux.go]: Can we only set the custom log path if we're not `--log-level=debug`? I want full logs from the OCI runtime in that case.

File: libpod/oci_linux.go [hunk 4]
```diff
@@ -187,6 +188,12 @@ func (r *OCIRuntime) conmonPackage() string {
 	return dpkgVersion(r.conmonPath)
 }
 
+type ociError struct {
```
Comment (mheon) [inline libpod/oci_linux.go]: This is thoroughly nonstandard and runc-specific, I imagine?

Can we get a comment of some sort to that effect?

File: libpod/oci_linux.go [hunk 5]
```diff
@@ -417,8 +427,16 @@ func (r *OCIRuntime) createOCIContainer(ctr *Container, cgroupParent string, res
 		}
 		logrus.Debugf("Received container pid: %d", ss.si.Pid)
 		if ss.si.Pid == -1 {
+			data, err := ioutil.ReadFile(ociLog)
+			if err == nil {
+				var ociErr ociError
+				if err := json.Unmarshal(data, &ociErr); err == nil {
+					return errors.Errorf("container create failed: %s", strings.Trim(ociErr.Msg, "\n"))
```
Comment (mheon) [inline libpod/oci_linux.go]: I would really like a wrapped error here so people can get a reasonable cause with errors.Cause() out of the API. Something like `ErrOCIRuntime` or `ErrRuntimeFailure`? If we don't want to print that in the error message we can discard the cause in Podman.

**Comment by giuseppe** [comment]
I've addressed the comments, pushed a new version.  I still think we should have a way to disable it somehow

**Comment by rhatdan** [comment]
@giuseppe Should we put something in the libpod.conf about how to call the OCI
oci flags or something like that?

**Comment by giuseppe** [comment]
> @giuseppe Should we put something in the libpod.conf about how to call the OCI
> oci flags or something like that?

the problem I had is that the flag will be global while we allow multiple OCI runtimes to be specified, can the TOML conf file be extended to have such a structured data?  Otherwise it won't really help when using different OCI runtimes, as you need to change the conf file every time.

**Comment by rhatdan** [comment]
I am fine with extending it.

File: libpod/oci_linux.go [hunk 6]
```diff
@@ -417,8 +423,16 @@ func (r *OCIRuntime) createOCIContainer(ctr *Container, cgroupParent string, res
...
+			if err == nil {
+				var ociErr ociError
+				if err := json.Unmarshal(data, &ociErr); err == nil {
+					return errors.Wrapf(ErrOCIRuntime, "%s", strings.Trim(ociErr.Msg, "\n"))
+				}
+			}
+			// If we failed to parse the JSON errors, then print the output as it is
 			if ss.si.Message != "" {
-				return errors.Wrapf(ErrInternal, "container create failed: %s", ss.si.Message)
+				return errors.Wrapf(ErrOCIRuntime, "%s", ss.si.Message)
```
Comment (TomSweeneyRedHat) [inline libpod/oci_linux.go]: Personally, I like having the "container created failed:" in the message.  Although I'd prefer it to be "container creation failed:".

Comment (TomSweeneyRedHat) [reply libpod/oci_linux.go]: LGTM, one comment for consideration.

**Comment by giuseppe** [comment]
added a new patch so that we can configure what runtimes support the json format

**Pull request merged**